### DISCUSSION TITLE
Does the ucrt compiler support the -fvisibility option?

### DISCUSSION THREAD

**Discussion by TheShermanTanker** [msys2/MINGW-packages#11945] [Q&A]
Title, the documentation online is incredibly sparse and unspecific as to what flags the ucrt compiler supports, and to make matters worse it also seems to just accept and silence some options silently without ever raising any issue that said flag doesn't have an effect whatsoever :/

**Comment by Biswa96** [comment]
According to this doc[^1], `-fvisibility` option is for ELF binaries only. The equivalent feature for Windows is dllexport and dllimport[^2] attributes.

[^1]: https://gcc.gnu.org/wiki/Visibility
[^2]: https://docs.microsoft.com/en-us/cpp/cpp/dllexport-dllimport

**Reply by TheShermanTanker** [reply]
Darn, was afraid of that. This has me thinking though, could we sift through all the options that MinGW-w64/UCRT gcc currently accepts that aren't relevant for compiling Windows binaries, and either properly adapt them for Windows or remove them entirely? For instance, `-fPIC` and `-fPIE` are definitively not relevant with the Windows PE format, yet are not only accepted and discarded silently, but actually _listed as options_ when `--help` is passed to gcc, and can even cause crashes if specified for either the linker or compiler, but not both. Other examples like rpath also come to mind, off the top of my head. On the other hand, you could also have visibility=hidden vs visibility=default serve the same purpose as dllexport does when on Windows, or maybe even act as an alias for it.

On a side note: After some testing UCRT gcc does seem to still follow the upstream gcc policy of exporting all symbols by default, unless a linker script is added to limit which symbols it exports, _or_ if dllexport is specified, whereby it then switches to only exporting items marked as dllexport. Weird...

**Reply by Biswa96** [reply]
I am not familiar with internals of gcc. Others many comment on the first question. But I think those options are just for compatibility reasons. Users have to take care of which options have any effect or not.

**Reply by mati865** [reply]
>  if dllexport is specified, whereby it then switches to only exporting items marked as dllexport.

This is upstream policy to somewhat match what MSVC does.
